### ISSUE TITLE
h5py fails to load in-memory hdf5 with the core driver 

### ISSUE THREAD

**Issue by HarveyYan** [h5py/h5py#2093]
To assist reproducing bugs, please include the following:
 * Operating System: Ubuntu 16.04.2 LTS
 * Python version: 3.6.8 or any 3.9.x
 * Where Python was acquired: anaconda
 * h5py version: 3.1.0/3.6.0
 * HDF5 version: 1.10.4/1.10.6
 
When I read a ~7.7GB hdf5 (with ~30000 groups, each representing an example in a dataset) in the normal mode (i.e. f=h5py.File('xxx.hdf5', 'r')), everything works fine (that is to say h5py can read the hdf5 file correctly and as expected).

When I load the **same** hdf5 directly into memory (e.g. f=h5py.File('xxx.hdf5', 'r', driver='core')), the file handle returns successfully but I cannot access any keys (e.g. f['some_key_that_exists']) without bumping some mysterious "RuntimeError: Link iteration failed (incorrect metadatda checksum after all read attempts)" error.

What is going on?

Additional detail: on win10 platform, using 3.9.12 python, 3.6.0 h5py with 1.10.6 hdf5 library, I can load in-memory hdf5 and successfully access any keys in it.

**Comment by ajelenak** [comment]
That error is coming from the HDF5 library. Can you enable the full HDF5 error stack by calling `h5py._errors.unsilence_errors()` before `f['some_key_that_exists']`?

You also seem to use older HDF5 versions: 1.10.4 and 1.10.6. I think h5py 3.6 comes with HDF5-1.12.1. And just last week 1.12.2 version was released. It would be very helpful if you could verify that your problem still exists with the latest version.

**Comment by HarveyYan** [comment]
Thanks for the quick repy!

Initally, I was using hdf5 group attributes to encode the content of a python dictionary which contains variable length numpy arrays. Later on, I somehow managed to dissect the python dictionary and save separately its keys and flattened arrays. So, the new hdf5 file only contains datasets of numpy arrays and the error goes away, i.e. I can successfully load and access in-memory hdf5 now.

I also attemped to try out hdf 1.12.(0/1/2) through anaconda, but somehow it gives me a version conflict so I stopped there.
